src/shmc/mod.rs

```diff
@@ -0,0 +1,508 @@
+//! have any notion of a round of computation: given the secrets `u` and `v`, every message related
+//! to multiplying `u` and `v` must be clearly distinguishable from any other message that is not
+//! related to `u` and `v`. It is possible to achieve by generating a unique identifier for every
+//! multiplication round and annotating messages that carry `u` or `v` information with it.
+//! Secondly, message processing and computation is done in the same thread and this implementation
+//! does not use async model which may be beneficial here for performance reasons.
+//! Lastly, using `u128` as a backing field to store shared secrets is too optimistic - a larger
+//! type is required to store secrets.
+//!
+//!
+//! [Paper] - <https://eprint.iacr.org/2019/1390.pdf>
+
+use hkdf::Hkdf;
+use rand::distributions::{Distribution, Standard};
+use rand::{thread_rng, Rng};
+use sha2::Sha256;
+
+/// Represents a secret in the system that must obey the following properties:
+/// - addition
+/// - multiplication
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct SharedSecret(u128);
+
+impl From<u128> for SharedSecret {
+    fn from(v: u128) -> Self {
+        SharedSecret(v)
+    }
+}
+
+impl SharedSecret {
+    /// Adds a given shared secret and returns a new instance
+    ///
+    /// ```
+    /// use raw_ipa::shmc::SharedSecret;
+    /// let u = SharedSecret::from(5);
+    /// let v = SharedSecret::from(10);
+    ///
+    /// assert_eq!(SharedSecret::from(15), u.wrapping_add(v));
+    /// ```
+    #[must_use]
+    pub fn wrapping_add(self, other: Self) -> Self {
+        SharedSecret(self.0.wrapping_add(other.0))
+    }
+
+    /// Subtracts a given shared secret and returns a new instance
+    ///
+    /// ```
+    /// use raw_ipa::shmc::SharedSecret;
+    /// let u = SharedSecret::from(3);
+    /// let v = SharedSecret::from(10);
+    ///
+    /// assert_eq!(SharedSecret::from(7), v.wrapping_sub(u));
+    /// ```
+    #[must_use]
+    pub fn wrapping_sub(self, other: Self) -> Self {
+        SharedSecret(self.0.wrapping_sub(other.0))
+    }
+
+    /// Multiplies by a given shared secret and returns a new instance
+    ///
+    /// ```
+    /// use raw_ipa::shmc::SharedSecret;
+    /// let u = SharedSecret::from(7);
+    /// let v = SharedSecret::from(11);
+    ///
+    /// assert_eq!(SharedSecret::from(77), v.wrapping_mul(u));
+    /// ```
+    #[must_use]
+    pub fn wrapping_mul(self, other: Self) -> Self {
+        SharedSecret(self.0.wrapping_mul(other.0))
+    }
+
+    /// Shares this secret to 3 parties by choosing three random elements (v1..v3)
+    /// under the constraint that v1+v2+v3 == self
+    #[must_use]
+    pub fn share(&self) -> [Share; 3] {
+        // thread_rng() is secure, but may not be secure enough for us. need to check
+        // what rand crate uses for it.
+        let u1 = thread_rng().gen::<Self>();
+        let u2 = thread_rng().gen::<Self>();
```

Comment:
Please don't internalize the RNG.  This is something that should be provided as an argument to this function.  I would also add an API that allows the caller to specify the random values explicitly, which can be invaluable in testing.  That means a signature like:

```rust
// type Rng = rand::Rng + rand::CryptoRng + ?Sized;
fn share(&self, rng: &mut impl Rng) -> [Share; 3] {
  self.share_with(rng.gen::<Self>(), rng.gen::<Self>())
}
fn share_with(&self, r1: Self::Type, r2: Self::Type) -> [Share; 3] {
  let r3 = self - r1 - r2;
  [Share(r1, r3), Share(r2, r1), Share(r3, r2)]
}
```

Reply:
+1 thanks! this is strictly better